[WIP] Allow passing preallocated segsbuf for alloc reuse

frankier opened the pull request:
This is a draft PR which allows allocation free usage of `quadgk(...)` by passing in an already allocated buffer. The use case is an application which has many 1D integrals repeated. This does seem to give a noticeable speedup.

Does this seem like something you might like to merge? Does the overall approach seem okay? If so I can clean this up (add docs and fix the segment type somehow).

stevengj commented on src/adapt.jl:
```diff
@@ -32,7 +35,8 @@ function do_quadgk(f::F, s::NTuple{N,T}, n, atol, rtol, maxevals, nrm) where {T,
         return (I, E) # fast return when no subdivisions required
```
A basic issue here is that it's nice to support allocation-free quadrature in cases when no subdivisions are required, which is the case when tuples are used as above.

stevengj commented on src/adapt.jl:
```diff
@@ -32,7 +35,8 @@ function do_quadgk(f::F, s::NTuple{N,T}, n, atol, rtol, maxevals, nrm) where {T,
     end
 
-    return adapt(f, heapify!(collect(segs), Reverse), I, E, numevals, x,w,gw,n, atol_, rtol_, maxevals, nrm)
+    heapify!(segbuf, Reverse)
```
Maybe instead, you can keep the above code the same (i.e., still use tuples), and then change this line to:
```
segheap = isnothing(segbuf) ? collect(segs) : (resize!(segbuf, N-1) .= segs)
heapify!(segheap, Reverse)
```
and then pass `segheap` to adapt

frankier replied:
Thanks for the feedback. Seems like a good idea and nice way of implementing it. Done!

frankier commented:
Done. I also added a helper function for allocating a segbuf, ``alloc_segbuf(...)` and some documentation. To get rid of some extra allocations I had to sprinkle `::F where {F}` in a couple more functions.

I added some tests for whether allocations actually occur. It does seem there are still a few allocations even in the no subdivision case. I could not quite figure out how to get rid of them all. From my perspective, this is "close enough" for now. It's just about possible that later versions of the Julia compiler will be able to elide some of these tiny allocations.

BTW the actual numbers from the test cases are:

```
[ Info: ("no_subdiv", 224)
[ Info: ("subdiv_alloc", 7584)
[ Info: ("subdiv_alloc_segbuf", 240)
```

frankier commented:
Should I update it for compatibility with Julia 0.7 and 1.0? I suppose the LTS is 1.6 now and these are both getting a bit old but if the plan is to retain compatibility I can at least get the tests to pass.

stevengj commented on src/adapt.jl:
```diff
@@ -32,11 +32,13 @@ function do_quadgk(f::F, s::NTuple{N,T}, n, atol, rtol, maxevals, nrm) where {T,
     end
 
-    return adapt(f, heapify!(collect(segs), Reverse), I, E, numevals, x,w,gw,n, atol_, rtol_, maxevals, nrm)
+    segheap = isnothing(segbuf) ? collect(segs) : (resize!(segbuf, N-1) .= segs)
```
```suggestion
    segheap = segbuf === nothing ? collect(segs) : (resize!(segbuf, N-1) .= segs)
```

stevengj commented:
It's not a big deal to drop support for Julia 1.0 (and 0.7 is basically the same), and we'll definitely do it at some point, but in this case it looks like just a 1-line fix so I wouldn't bother.

stevengj commented on test/runtests.jl:
```diff
@@ -94,4 +94,28 @@ end
+    segbuf = alloc_segbuf(size=1)
+    subdiv_alloc_segbuf() = @timed quadgk(osc, -1.0, 1.0, segbuf=segbuf)
+    no_subdiv() # warmup
+    @test no_subdiv().bytes < smallallocbytes
+    subdiv_alloc() # warmup
+    @test subdiv_alloc().bytes > smallallocbytes
+    subdiv_alloc_segbuf() # warmup
+    @test subdiv_alloc_segbuf().bytes < smallallocbytes
```
```suggestion
    @test no_subdiv()[3] < smallallocbytes # [3] == .bytes starting in Julia 1.5
    subdiv_alloc() # warmup
    @test subdiv_alloc()[3] > smallallocbytes
    subdiv_alloc_segbuf() # warmup
    @test subdiv_alloc_segbuf()[3] < smallallocbytes
```

stevengj commented on src/adapt.jl:
```diff
@@ -169,15 +171,35 @@ or `1/sqrt(x)` singularity).
+
+In normal usage, `quadgk(...)` will allocate a buffer for segments. You can
+instead pass a preallocated buffer allocated using `alloc_segbuf(...)` as the
+`segbuf` argument. This buffer can be used across muliple calls to avoid
```
```suggestion
`segbuf` argument. This buffer can be used across multiple calls to avoid
```

frankier commented:
Thanks!
